File: code/Java/Chapter12/12-5/Knapsack.java
public class Knapsack {
    public static class Item {
        double weight;
        double value;

        public Item(double weight, double value) {
            this.weight = weight;
            this.value = value;
        }
    }

    public static void main(String[] args) {
        Item[] items = {
                new Item(2, 3),
                new Item(3, 4),
                new Item(4, 5),
                new Item(5, 8),
                new Item(9, 10)
        };
        int capacity = 20;
        System.out.println(knapsack(items, capacity));
    }

    public static int knapsack(Item[] items, int capacity) {
        int[][] values = new int[items.length + 1][capacity + 1];

        for (int i = 1; i <= items.length; i++) {
            for (int j = 1; j <= capacity; j++) {
                if (items[i - 1].weight > j) {
                    values[i][j] = values[i - 1][j];
                } else {
                    values[i][j] = Math.max(
                            values[i - 1][j - (int) items[i - 1].weight] + (int) items[i - 1].value,
                            values[i - 1][j]
                    );
                }
            }
        }

        return values[items.length][capacity];
    }
}

File: code/C++/Chapter12/12-5/knapsack.cpp
#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

class Item {
    public:
    int weight;
    int value;

    Item(int weight, int value) {
        this->weight = weight;
        this->value = value;
    }
};

int knapsack(vector<Item> items, int capacity) {
    int values[items.size() + 1][capacity + 1];
    memset(values, 0, sizeof(values));

    for (int i = 1; i <= (int)items.size(); i++) {
        for (int j = 1; j <= capacity; j++) {
            if (items[i - 1].weight > j) {
                values[i][j] = values[i - 1][j];
            } else {
                values[i][j] = max(
                    values[i - 1][j - items[i - 1].weight] + items[i - 1].value,
                    values[i - 1][j]
                );
            }
        }
    }

    return values[items.size()][capacity];
}

int main() {
    vector<Item> items = {
        Item(2, 3),
        Item(3, 4),
        Item(4, 5),
        Item(5, 8),
        Item(9, 10),
    };
    int capacity = 20;

    cout << knapsack(items, capacity) << endl;
    return 0;
}